{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "####import needed modules\n",
    "\n",
    "from IPython import get_ipython\n",
    "get_ipython().run_line_magic('matplotlib', 'qt')\n",
    "import pickle\n",
    "from appJar import gui\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Get the dataframe of the results###\n",
    "file0 = open('Human_Means','rb')\n",
    "file = open('medians_df', 'rb')\n",
    "meds = pickle.load(file)\n",
    "file.close()\n",
    "human_means = pickle.load(file0)\n",
    "file0.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Get ndArray of all column names : MOUSE\n",
    "columnsNamesArr = meds.columns.values\n",
    "listOfColumnNames = list(columnsNamesArr)\n",
    "idx = list(meds.index)\n",
    "\n",
    "cellList = ['Lamp5', 'Lamp5 Lhx6', 'Snc','Sncg/Ndnf',\n",
    "            'Serpfin','Vip','Sst','Sst Chodl','Pvalb','PIR six',\n",
    "            'DG','RSP/ACA L4/5 IT','L2/3 IT Cxcl14','L2/3 IT Cdc14a',\n",
    "            'IT RHP Dcn','L2/3 IT Otof','L6 IT','L4 IT','L4/5 IT',\n",
    "            'L5 IT','RSP_ACA IT Scnn1a', 'Car3','L5 ET','SUB-Sp Ndst4',\n",
    "            'NP','Ly6g6e','RHP Cplx3','L6 CT','L6b','CA1sp/SUB-sp Kcnip1',\n",
    "            'CA1sp','CA3sp','CR','Astro','Oligo','VLMC','Endo','SMC',\n",
    "            'Macrophage']\n",
    "\n",
    "\n",
    "# Get ndArray of all column names: HUMAN\n",
    "human_columnsNamesArr = human_means.columns.values\n",
    "human_listOfColumnNames = list(human_columnsNamesArr)\n",
    "human_idx = list(human_means.index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-04-30 11:58:35,568 appJar:WARNING [799/_checkMode]: Interactive mode is not fully tested, some features might not work.\n",
      "2020-04-30 11:58:35,580 appJar:WARNING [810/_checkMode]: iPython is not fully tested, some features might not work.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#############choose analysis parameters##################\n",
    "Outputs = {}\n",
    "\n",
    "def excel(excel):\n",
    "    return(excel)\n",
    "\n",
    "def launch(win):\n",
    "    app.showSubWindow(win, hide = True)\n",
    "    app.hideSubWindow(\"Starter\")\n",
    "\n",
    "def glaunch(win):\n",
    "    app.showSubWindow(win, hide = True)\n",
    "    if win == \"Mouse Genes\":\n",
    "        app.hideSubWindow(\"Mouse\")\n",
    "        Outputs['Cell Types'] = app.getListBox('mouse_cells')\n",
    "    if win == \"Human Genes\":\n",
    "        app.hideSubWindow(\"Human\")\n",
    "        Outputs['Cell Types'] = app.getListBox('human_cells')\n",
    "\n",
    "def press(btn):\n",
    "    app.stop()\n",
    "       \n",
    "def press_go(btn):\n",
    "    if btn == 'Submit Human':\n",
    "        geneslist =[]\n",
    "        geneslist.append(app.getEntry(\" Choose Gene 1\"))\n",
    "        geneslist.append(app.getEntry(\" Choose Gene 2\"))\n",
    "        geneslist.append(app.getEntry(\" Choose Gene 3\"))\n",
    "        geneslist.append(app.getEntry(\" Choose Gene 4\"))\n",
    "        geneslist.append(app.getEntry(\" Choose Gene 5\"))\n",
    "        geneslist.append(app.getEntry(\" Choose Gene 6\")) \n",
    "    elif btn == 'Submit Mouse':\n",
    "        geneslist =[]\n",
    "        geneslist.append(app.getEntry(\"Choose Gene 1 \"))\n",
    "        geneslist.append(app.getEntry(\"Choose Gene 2 \"))\n",
    "        geneslist.append(app.getEntry(\"Choose Gene 3 \"))\n",
    "        geneslist.append(app.getEntry(\"Choose Gene 4 \"))\n",
    "        geneslist.append(app.getEntry(\"Choose Gene 5 \"))\n",
    "        geneslist.append(app.getEntry(\"Choose Gene 6 \")) \n",
    "    Outputs['Genes'] = geneslist\n",
    "    Outputs['Species'] = str(btn)\n",
    "    app.stop()\n",
    "\n",
    "\n",
    "app = gui()\n",
    "app.startSubWindow(\"Starter\")\n",
    "# ##Initial gui\n",
    "app.setStretch(\"both\")\n",
    "app.setOnTop(stay=True)\n",
    "app.setFont(12)\n",
    "app.addLabel(\"l0\", \"Welcome: Browse the Allen Transcriptomics single cell RNAseq\", 0,0,4)\n",
    "app.addLabel(\"spacer1\", \" \")\n",
    "app.addLabel(\"spacer2\", \" \")\n",
    "app.addLabel(\"l0a\", \"Human, Mouse or Both?\", 1,0,4)\n",
    "app.setLabelBg(\"l0\", \"purple\")\n",
    "app.setLabelBg(\"l0a\", \"purple\")\n",
    "app.setLabelFg(\"l0\", \"white\")\n",
    "app.setLabelFg(\"l0a\", \"white\")\n",
    "app.addButtons([\"Human\", \"Mouse\"], launch)\n",
    "app.addCheckBox(\"Create excel file? If so, check box and name it below\")\n",
    "app.addEntry(\"Excel Name\")\n",
    "app.addButton(\"Cancel\", press)\n",
    "app.stopSubWindow()\n",
    "app.showSubWindow(\"Starter\", hide = True)\n",
    "\n",
    "#################Gene Choice Windows\n",
    "\n",
    "####define genes window for \"Human\"\n",
    "app.startSubWindow(\"Human Genes\")\n",
    "app.addLabel(\"l2c\", \" Choose Genes\", 4,0,4,0, \"white\")\n",
    "app.setLabelBg(\"l2c\", \"purple\")\n",
    "app.setLabelFg(\"l2c\", \"white\")\n",
    "app.addLabel(\"spaceragainc\", \" \",5,0,4)\n",
    "app.addLabel(\"title1c\", \"Gene 1:\")\n",
    "app.addAutoEntry(\" Choose Gene 1\", human_idx)\n",
    "app.setAutoEntryNumRows(\" Choose Gene 1\", 15)\n",
    "app.addLabel(\"title2c\", \"Gene 2:\")\n",
    "app.addAutoEntry(\" Choose Gene 2\", human_idx)\n",
    "app.setAutoEntryNumRows(\" Choose Gene 2\", 15)\n",
    "app.addLabel(\"title3c\", \"Gene 3:\")\n",
    "app.addAutoEntry(\" Choose Gene 3\", human_idx)\n",
    "app.setAutoEntryNumRows(\" Choose Gene 3\", 15)\n",
    "app.addLabel(\"title4c\", \"Gene 4:\")\n",
    "app.addAutoEntry(\" Choose Gene 4\", human_idx)\n",
    "app.setAutoEntryNumRows(\" Choose Gene 4\", 15)\n",
    "app.addLabel(\"title5c\", \"Gene 5:\")\n",
    "app.addAutoEntry(\" Choose Gene 5\", human_idx)\n",
    "app.setAutoEntryNumRows(\" Choose Gene 5\", 15)\n",
    "app.addLabel(\"title6c\", \"Gene 6:\")\n",
    "app.addAutoEntry(\" Choose Gene 6\", human_idx)\n",
    "app.setAutoEntryNumRows(\" Choose Gene 6\", 15)\n",
    "app.addLabel(\"spacer9c\", \" \")\n",
    "app.addButton(\"Submit Human\", press_go)\n",
    "app.addButton(\"Cancel Human\", press)\n",
    "app.addLabel(\"spacer0c\", \" \")\n",
    "app.stopSubWindow()\n",
    "\n",
    "####define genes window for \"Mouse\"\n",
    "app.startSubWindow(\"Mouse Genes\")\n",
    "app.addLabel(\"l2d\", \"Choose Genes \", 4,0,4,0, \"white\")\n",
    "app.setLabelBg(\"l2d\", \"purple\")\n",
    "app.setLabelFg(\"l2d\", \"white\")\n",
    "app.addLabel(\"spaceragaind\", \" \",5,0,4)\n",
    "app.addLabel(\"title1d\", \"Gene 1:\")\n",
    "app.addAutoEntry(\"Choose Gene 1 \", idx)\n",
    "app.setAutoEntryNumRows(\"Choose Gene 1 \", 15)\n",
    "app.addLabel(\"title2d\", \"Gene 2:\")\n",
    "app.addAutoEntry(\"Choose Gene 2 \", idx)\n",
    "app.setAutoEntryNumRows(\"Choose Gene 2 \", 15)\n",
    "app.addLabel(\"title3d\", \"Gene 3:\")\n",
    "app.addAutoEntry(\"Choose Gene 3 \", idx)\n",
    "app.setAutoEntryNumRows(\"Choose Gene 3 \", 15)\n",
    "app.addLabel(\"title4d\", \"Gene 4:\")\n",
    "app.addAutoEntry(\"Choose Gene 4 \", idx)\n",
    "app.setAutoEntryNumRows(\"Choose Gene 4 \", 15)\n",
    "app.addLabel(\"title5d\", \"Gene 5:\")\n",
    "app.addAutoEntry(\"Choose Gene 5 \", idx)\n",
    "app.setAutoEntryNumRows(\"Choose Gene 5 \", 15)\n",
    "app.addLabel(\"title6d\", \"Gene 6:\")\n",
    "app.addAutoEntry(\"Choose Gene 6 \", idx)\n",
    "app.setAutoEntryNumRows(\"Choose Gene 6 \", 15)\n",
    "app.addLabel(\"spacer9d\", \" \")\n",
    "app.addButton(\"Submit Mouse\", press_go)\n",
    "app.addButton(\"Cancel Mouse\", press)\n",
    "app.addLabel(\"spacer0d\", \" \")\n",
    "app.stopSubWindow()\n",
    "\n",
    "#########################cell types windows######################################\n",
    "###define human cell types window\n",
    "app.startSubWindow(\"Human\", modal=True)\n",
    "app.setOnTop(stay=True)\n",
    "app.addLabel(\"l1\", \"Select Human Cell Types:\", 0,0,1)\n",
    "app.addListBox(\"human_cells\", human_listOfColumnNames)\n",
    "app.setListBoxMulti(\"human_cells\", multi=True)\n",
    "app.addHorizontalSeparator(3,0,4)\n",
    "app.addButton(\"Human Genes\", glaunch)\n",
    "app.addButton(\" Cancel \", press)\n",
    "app.stopSubWindow()\n",
    "###define mouse cell types window\n",
    "app.startSubWindow(\"Mouse\", modal=True)\n",
    "app.setOnTop(stay=True)\n",
    "app.addLabel(\"l2\", \"Select Mouse Cell Types:\", 0,0,1)\n",
    "app.addListBox(\"mouse_cells\", cellList)\n",
    "app.setListBoxMulti(\"mouse_cells\", multi=True)\n",
    "app.addHorizontalSeparator(3,0,4)\n",
    "app.addButton(\"Mouse Genes\", glaunch)\n",
    "app.addButton(\"Cancel \", press)\n",
    "app.stopSubWindow()\n",
    "\n",
    "app.go()\n",
    "###########################################################3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\dbaka\\Anaconda3\\lib\\site-packages\\ipykernel_launcher.py:45: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#########################################################\n",
    "Excel = app.getCheckBox(\"Create excel file? If so, check box and name it below\")\n",
    "\n",
    "##clean up gene list in case you chose less than 6 genes#######\n",
    "Outputs['Genes'] = list(filter(None, Outputs['Genes']))\n",
    "\n",
    "####split the medians data by cell and gene selection############\n",
    "from functools import partial, reduce \n",
    "GeneByRegion ={}\n",
    "myRegion = {}\n",
    "\n",
    "if  Outputs['Species'] == \"Submit Mouse\":\n",
    "    for p in Outputs['Cell Types']:\n",
    "        myRegion[p] = meds.filter(regex=p)\n",
    "        myRegion[p]  = myRegion[p].loc[Outputs['Genes']]\n",
    "        m_name = \"mean_\" + p\n",
    "        sem_name = \"sem_\" + p\n",
    "        regtemp = myRegion[p]\n",
    "        regtemp[m_name] = myRegion[p].mean(axis=1)\n",
    "        regtemp[sem_name] = myRegion[p].sem(axis=1)\n",
    "        my_reduce = partial(pd.merge, on='GeneID', how='outer')                                                              \n",
    "        output = reduce(my_reduce, myRegion.values())\n",
    "        GeneByRegion[p] = output.iloc[:,-2]\n",
    "    Outputs['means'] = output[[col for col in output.columns if \"mean\" in col]].reset_index().set_index('GeneID')\n",
    "    # Outputs['sems'] = output[[col for col in output.columns if \"sem\" in col]].reset_index().set_index('GeneID')\n",
    "    # # from toolz import interleave\n",
    "    # # mean_and_sem = pd.concat([means, sems], axis=1)[list(interleave([means, sems]))]\n",
    "                      \n",
    "if  Outputs['Species'] == \"Submit Human\":\n",
    "    for p in Outputs['Cell Types']:  \n",
    "        myRegion[p] = human_means.filter(regex=p)\n",
    "        myRegion[p]  = myRegion[p].loc[Outputs['Genes']]\n",
    "        my_reduce = partial(pd.merge, on='GeneID', how='outer')                                                              \n",
    "        output = reduce(my_reduce, myRegion.values())\n",
    "        GeneByRegion[p] = output.iloc[:,-2]\n",
    "    Outputs['means'] = output[[col for col in output.columns]].reset_index().set_index('GeneID')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "##make a seaborn catplot from it - make the dict into a dataframe and work from that\n",
    "import seaborn as sns\n",
    "plotgenes = pd.DataFrame.from_dict(GeneByRegion).reset_index()\n",
    "genesarray0 = plotgenes['GeneID']\n",
    "genesarray = pd.Series()\n",
    "for i in range(len(Outputs[\"Cell Types\"])):\n",
    "    genesarray = genesarray.append(genesarray0)\n",
    "genesarray=genesarray.tolist()    \n",
    "df2 = pd.concat([\n",
    "    plotgenes[Outputs['Cell Types']].melt(value_name='Expression Level')], \n",
    "    axis=1)\n",
    "df2['Gene'] = genesarray\n",
    "maxm = max(df2[\"Expression Level\"])\n",
    "\n",
    "\n",
    "###Now graph it####################################################\n",
    "    \n",
    "sns.set_palette(\"Blues\")\n",
    "g = sns.catplot(x=\"Gene\", y=\"Expression Level\", col=\"variable\",\n",
    "                data=df2, saturation=.9,\n",
    "                kind=\"bar\", ci=None, aspect=.6, edgecolor=\"k\", linewidth=1)\n",
    "(g.set_axis_labels(\"\", \"Expression\")\n",
    "  .set_xticklabels(Outputs[\"Genes\"])\n",
    "  .set_titles(\"{col_name}\")\n",
    "  .set(ylim=(0, (maxm + 1)))\n",
    "  .despine(left=True))  \n",
    "\n",
    "for ax in g.axes.ravel():\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=45, horizontalalignment='right')\n",
    "    for p in ax.patches:\n",
    "        ax.annotate('%.2f' % p.get_height(), (p.get_x() + p.get_width() / 2., p.get_height()), \n",
    "\n",
    "                    ha='center', va='center', xytext=(0, 8), \n",
    "                    textcoords='offset points')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Create an excel file ##################\n",
    "if Excel == True:\n",
    "    FileName = app.getEntry(\"Excel Name\")\n",
    "    FileName = FileName + \".xlsx\"\n",
    "    df2.to_excel(FileName) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
