{"cells":[{"metadata":{"trusted":true},"cell_type":"code","source":"pip install pandas","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"pip install appjar","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"pip install matplotlib","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"pip install seaborn","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"\n####import needed modules\n\nimport pickle\nfrom appJar import gui\nimport pandas as pd\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import zipfile\n\nzf = zipfile.ZipFile('meds.zip') \nmeds = pd.read_csv(zf.open('meds.csv')).set_index('GeneID')\nprint(meds.head())\nzf2 = zipfile.ZipFile('human_means.zip')\nhuman_means = pd.read_csv(zf2.open('human_means.csv')).set_index('GeneID')\nprint(human_means.head())","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# Get ndArray of all column names : MOUSE\ncolumnsNamesArr = meds.columns.values\nlistOfColumnNames = list(columnsNamesArr)\nidx = list(meds.index)\n\ncellList = ['Lamp5', 'Lamp5 Lhx6', 'Snc','Sncg/Ndnf',\n            'Serpfin','Vip','Sst','Sst Chodl','Pvalb','PIR six',\n            'DG','RSP/ACA L4/5 IT','L2/3 IT Cxcl14','L2/3 IT Cdc14a',\n            'IT RHP Dcn','L2/3 IT Otof','L6 IT','L4 IT','L4/5 IT',\n            'L5 IT','RSP_ACA IT Scnn1a', 'Car3','L5 ET','SUB-Sp Ndst4',\n            'NP','Ly6g6e','RHP Cplx3','L6 CT','L6b','CA1sp/SUB-sp Kcnip1',\n            'CA1sp','CA3sp','CR','Astro','Oligo','VLMC','Endo','SMC',\n            'Macrophage']\n\n\n# Get ndArray of all column names: HUMAN\nhuman_columnsNamesArr = human_means.columns.values\nhuman_listOfColumnNames = list(human_columnsNamesArr)\nhuman_idx = list(human_means.index)\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"\n#############choose analysis parameters##################\nOutputs = {}\n\ndef excel(excel):\n    return(excel)\n\ndef launch(win):\n    app.showSubWindow(win, hide = True)\n    app.hideSubWindow(\"Starter\")\n\ndef glaunch(win):\n    app.showSubWindow(win, hide = True)\n    if win == \"Mouse Genes\":\n        app.hideSubWindow(\"Mouse\")\n        Outputs['Cell Types'] = app.getListBox('mouse_cells')\n    if win == \"Human Genes\":\n        app.hideSubWindow(\"Human\")\n        Outputs['Cell Types'] = app.getListBox('human_cells')\n\ndef press(btn):\n    app.stop()\n       \ndef press_go(btn):\n    if btn == 'Submit Human':\n        geneslist =[]\n        geneslist.append(app.getEntry(\" Choose Gene 1\"))\n        geneslist.append(app.getEntry(\" Choose Gene 2\"))\n        geneslist.append(app.getEntry(\" Choose Gene 3\"))\n        geneslist.append(app.getEntry(\" Choose Gene 4\"))\n        geneslist.append(app.getEntry(\" Choose Gene 5\"))\n        geneslist.append(app.getEntry(\" Choose Gene 6\")) \n    elif btn == 'Submit Mouse':\n        geneslist =[]\n        geneslist.append(app.getEntry(\"Choose Gene 1 \"))\n        geneslist.append(app.getEntry(\"Choose Gene 2 \"))\n        geneslist.append(app.getEntry(\"Choose Gene 3 \"))\n        geneslist.append(app.getEntry(\"Choose Gene 4 \"))\n        geneslist.append(app.getEntry(\"Choose Gene 5 \"))\n        geneslist.append(app.getEntry(\"Choose Gene 6 \")) \n    Outputs['Genes'] = geneslist\n    Outputs['Species'] = str(btn)\n    app.stop()\n\n\napp = gui()\napp.startSubWindow(\"Starter\")\n# ##Initial gui\napp.setStretch(\"both\")\napp.setOnTop(stay=True)\napp.setFont(12)\napp.addLabel(\"l0\", \"Welcome: Browse the Allen Transcriptomics single cell RNAseq\", 0,0,4)\napp.addLabel(\"spacer1\", \" \")\napp.addLabel(\"spacer2\", \" \")\napp.addLabel(\"l0a\", \"Human, Mouse or Both?\", 1,0,4)\napp.setLabelBg(\"l0\", \"purple\")\napp.setLabelBg(\"l0a\", \"purple\")\napp.setLabelFg(\"l0\", \"white\")\napp.setLabelFg(\"l0a\", \"white\")\napp.addButtons([\"Human\", \"Mouse\"], launch)\napp.addCheckBox(\"Create excel file? If so, check box and name it below\")\napp.addEntry(\"Excel Name\")\napp.addButton(\"Cancel\", press)\napp.stopSubWindow()\napp.showSubWindow(\"Starter\", hide = True)\n\n#################Gene Choice Windows\n\n####define genes window for \"Human\"\napp.startSubWindow(\"Human Genes\")\napp.addLabel(\"l2c\", \" Choose Genes\", 4,0,4,0, \"white\")\napp.setLabelBg(\"l2c\", \"purple\")\napp.setLabelFg(\"l2c\", \"white\")\napp.addLabel(\"spaceragainc\", \" \",5,0,4)\napp.addLabel(\"title1c\", \"Gene 1:\")\napp.addAutoEntry(\" Choose Gene 1\", human_idx)\napp.setAutoEntryNumRows(\" Choose Gene 1\", 15)\napp.addLabel(\"title2c\", \"Gene 2:\")\napp.addAutoEntry(\" Choose Gene 2\", human_idx)\napp.setAutoEntryNumRows(\" Choose Gene 2\", 15)\napp.addLabel(\"title3c\", \"Gene 3:\")\napp.addAutoEntry(\" Choose Gene 3\", human_idx)\napp.setAutoEntryNumRows(\" Choose Gene 3\", 15)\napp.addLabel(\"title4c\", \"Gene 4:\")\napp.addAutoEntry(\" Choose Gene 4\", human_idx)\napp.setAutoEntryNumRows(\" Choose Gene 4\", 15)\napp.addLabel(\"title5c\", \"Gene 5:\")\napp.addAutoEntry(\" Choose Gene 5\", human_idx)\napp.setAutoEntryNumRows(\" Choose Gene 5\", 15)\napp.addLabel(\"title6c\", \"Gene 6:\")\napp.addAutoEntry(\" Choose Gene 6\", human_idx)\napp.setAutoEntryNumRows(\" Choose Gene 6\", 15)\napp.addLabel(\"spacer9c\", \" \")\napp.addButton(\"Submit Human\", press_go)\napp.addButton(\"Cancel Human\", press)\napp.addLabel(\"spacer0c\", \" \")\napp.stopSubWindow()\n\n####define genes window for \"Mouse\"\napp.startSubWindow(\"Mouse Genes\")\napp.addLabel(\"l2d\", \"Choose Genes \", 4,0,4,0, \"white\")\napp.setLabelBg(\"l2d\", \"purple\")\napp.setLabelFg(\"l2d\", \"white\")\napp.addLabel(\"spaceragaind\", \" \",5,0,4)\napp.addLabel(\"title1d\", \"Gene 1:\")\napp.addAutoEntry(\"Choose Gene 1 \", idx)\napp.setAutoEntryNumRows(\"Choose Gene 1 \", 15)\napp.addLabel(\"title2d\", \"Gene 2:\")\napp.addAutoEntry(\"Choose Gene 2 \", idx)\napp.setAutoEntryNumRows(\"Choose Gene 2 \", 15)\napp.addLabel(\"title3d\", \"Gene 3:\")\napp.addAutoEntry(\"Choose Gene 3 \", idx)\napp.setAutoEntryNumRows(\"Choose Gene 3 \", 15)\napp.addLabel(\"title4d\", \"Gene 4:\")\napp.addAutoEntry(\"Choose Gene 4 \", idx)\napp.setAutoEntryNumRows(\"Choose Gene 4 \", 15)\napp.addLabel(\"title5d\", \"Gene 5:\")\napp.addAutoEntry(\"Choose Gene 5 \", idx)\napp.setAutoEntryNumRows(\"Choose Gene 5 \", 15)\napp.addLabel(\"title6d\", \"Gene 6:\")\napp.addAutoEntry(\"Choose Gene 6 \", idx)\napp.setAutoEntryNumRows(\"Choose Gene 6 \", 15)\napp.addLabel(\"spacer9d\", \" \")\napp.addButton(\"Submit Mouse\", press_go)\napp.addButton(\"Cancel Mouse\", press)\napp.addLabel(\"spacer0d\", \" \")\napp.stopSubWindow()\n\n#########################cell types windows######################################\n###define human cell types window\napp.startSubWindow(\"Human\", modal=True)\napp.setOnTop(stay=True)\napp.addLabel(\"l1\", \"Select Human Cell Types:\", 0,0,1)\napp.addListBox(\"human_cells\", human_listOfColumnNames)\napp.setListBoxMulti(\"human_cells\", multi=True)\napp.addHorizontalSeparator(3,0,4)\napp.addButton(\"Human Genes\", glaunch)\napp.addButton(\" Cancel \", press)\napp.stopSubWindow()\n###define mouse cell types window\napp.startSubWindow(\"Mouse\", modal=True)\napp.setOnTop(stay=True)\napp.addLabel(\"l2\", \"Select Mouse Cell Types:\", 0,0,1)\napp.addListBox(\"mouse_cells\", cellList)\napp.setListBoxMulti(\"mouse_cells\", multi=True)\napp.addHorizontalSeparator(3,0,4)\napp.addButton(\"Mouse Genes\", glaunch)\napp.addButton(\"Cancel \", press)\napp.stopSubWindow()\n\napp.go()\n###########################################################3\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"\n#########################################################\nExcel = app.getCheckBox(\"Create excel file? If so, check box and name it below\")\n\n##clean up gene list in case you chose less than 6 genes#######\nOutputs['Genes'] = list(filter(None, Outputs['Genes']))\n\n####split the medians data by cell and gene selection############\nfrom functools import partial, reduce \nGeneByRegion ={}\nmyRegion = {}\n\nif  Outputs['Species'] == \"Submit Mouse\":\n    for p in Outputs['Cell Types']:\n        myRegion[p] = meds.filter(regex=p)\n        myRegion[p]  = myRegion[p].loc[Outputs['Genes']]\n        m_name = \"mean_\" + p\n        sem_name = \"sem_\" + p\n        regtemp = myRegion[p]\n        regtemp[m_name] = myRegion[p].mean(axis=1)\n        regtemp[sem_name] = myRegion[p].sem(axis=1)\n        my_reduce = partial(pd.merge, on='GeneID', how='outer')                                                              \n        output = reduce(my_reduce, myRegion.values())\n        GeneByRegion[p] = output.iloc[:,-2]\n    Outputs['means'] = output[[col for col in output.columns if \"mean\" in col]].reset_index().set_index('GeneID')\n    # Outputs['sems'] = output[[col for col in output.columns if \"sem\" in col]].reset_index().set_index('GeneID')\n    # # from toolz import interleave\n    # # mean_and_sem = pd.concat([means, sems], axis=1)[list(interleave([means, sems]))]\n                      \nif  Outputs['Species'] == \"Submit Human\":\n    for p in Outputs['Cell Types']:  \n        myRegion[p] = human_means.filter(regex=p)\n        myRegion[p]  = myRegion[p].loc[Outputs['Genes']]\n        my_reduce = partial(pd.merge, on='GeneID', how='outer')                                                              \n        output = reduce(my_reduce, myRegion.values())\n        GeneByRegion[p] = output.iloc[:,-2]\n    Outputs['means'] = output[[col for col in output.columns]].reset_index().set_index('GeneID')\n\n\n\n\n##make a seaborn catplot from it - make the dict into a dataframe and work from that\nimport seaborn as sns\nplotgenes = pd.DataFrame.from_dict(GeneByRegion).reset_index()\ngenesarray0 = plotgenes['GeneID']\ngenesarray = pd.Series()\nfor i in range(len(Outputs[\"Cell Types\"])):\n    genesarray = genesarray.append(genesarray0)\ngenesarray=genesarray.tolist()    \ndf2 = pd.concat([\n    plotgenes[Outputs['Cell Types']].melt(value_name='Expression Level')], \n    axis=1)\ndf2['Gene'] = genesarray\nmaxm = max(df2[\"Expression Level\"])\n\n\n###Now graph it####################################################\n    \nsns.set_palette(\"Blues\")\ng = sns.catplot(x=\"Gene\", y=\"Expression Level\", col=\"variable\",\n                data=df2, saturation=.9,\n                kind=\"bar\", ci=None, aspect=.6, edgecolor=\"k\", linewidth=1)\n(g.set_axis_labels(\"\", \"Expression\")\n  .set_xticklabels(Outputs[\"Genes\"])\n  .set_titles(\"{col_name}\")\n  .set(ylim=(0, (maxm + 1)))\n  .despine(left=True))  \n\nfor ax in g.axes.ravel():\n    ax.set_xticklabels(ax.get_xticklabels(), rotation=45, horizontalalignment='right')\n    for p in ax.patches:\n        ax.annotate('%.2f' % p.get_height(), (p.get_x() + p.get_width() / 2., p.get_height()), \n\n                    ha='center', va='center', xytext=(0, 8), \n                    textcoords='offset points')\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"###Create an excel file ##################\nif Excel == True:\n    FileName = app.getEntry(\"Excel Name\")\n    FileName = FileName + \".xlsx\"\n    df2.to_excel(FileName) ","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"name":"python3","display_name":"Python 3","language":"python"},"language_info":{"name":"python","version":"3.7.6","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat":4,"nbformat_minor":4}