{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "hideCode": false,
    "hidePrompt": false
   },
   "source": [
    "# Press the Double Arrow above to Begin, then click red pop-up to confirm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hideCode": false,
    "hidePrompt": false,
    "hide_input": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "display(\"Loading Allen Transcriptomics Single-Cell RNAseq data, retreived May 2020. Be patient, it's loading!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "hideCode": false,
    "hidePrompt": false,
    "hide_input": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".output_wrapper, .output {\n",
       "    height:auto !important;\n",
       "    max-height:1000px;  /* your desired max-height here */\n",
       "}\n",
       ".output_scroll {\n",
       "    box-shadow:none !important;\n",
       "    webkit-box-shadow:none !important;\n",
       "}\n",
       "</style>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%html\n",
    "<style>\n",
    ".output_wrapper, .output {\n",
    "    height:auto !important;\n",
    "    max-height:1000px;  /* your desired max-height here */\n",
    "}\n",
    ".output_scroll {\n",
    "    box-shadow:none !important;\n",
    "    webkit-box-shadow:none !important;\n",
    "}\n",
    "</style>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "hideCode": false,
    "hidePrompt": false,
    "hide_input": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "import ipywidgets as widgets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": [
    "\n",
    "mouse_means0 = pd.read_csv('https://idk-etl-prod-download-bucket.s3.amazonaws.com/aibs_mouse_ctx-hip_smart-seq/trimmed_means.csv')\n",
    "human_means0 = pd.read_csv('https://idk-etl-prod-download-bucket.s3.amazonaws.com/aibs_human_ctx_smart-seq/trimmed_means.csv')\n",
    "\n",
    "# Get all gene and cell type names: HUMAN\n",
    "human_idx = list(human_means0.columns.values)\n",
    "human_listOfColumnNames = list(human_means0['cluster_label'])\n",
    "# Get all gene and cell type: MOUSE\n",
    "mouse_idx = list(mouse_means0.columns.values)\n",
    "mouse_listOfColumnNames = list(mouse_means0['cluster_label'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#combine clusters into higher_clusters\n",
    "newlist =[]\n",
    "for i in mouse_listOfColumnNames:\n",
    "    i = str(i)\n",
    "    head, sep, tail = i.partition('_') \n",
    "    newlist.append(tail) \n",
    "mouse_means0['higher_cluster'] = newlist\n",
    "\n",
    "\n",
    "\n",
    "#list what cell clusters count as which cell types\n",
    "excite_list = ['CA1','CA1-ProS','CA2-IG-FC','CA3','CT SUB','DG', 'L2 IT ENTl',\n",
    " 'L2 IT ENTm','L2 IT HATA','L2 IT PAR','L2 IT ProS','L2 IT RSP-ACA','L2 IT RSPv',\n",
    " 'L2/3 IT APr','L2/3 IT CTX','L2/3 IT ENTl','L2/3 IT PPP','L2/3 IT TPE',\n",
    " 'L3 IT ENTl','L3 IT ENTm','L3 RSP-ACA','L4/5 IT CTX','L4/5 IT TPE-ENT',\n",
    " 'L5 IT CTX','L5 IT TPE-ENT','L5 NP CTX','L5 PPP','L5 PT CTX','L5 PT RSP-ACA',\n",
    " 'L5/6 IT CTX', 'L6 Car3', 'L6 CT CTX','L6 CT ENTm','L6 IT CTX','L6 IT ENTl',\n",
    " 'L6 IT RHP','L6 NP CT CTX','L6b CTX','L6b RHP','L6b/CT ENT','Pax6','SUB','SUB-ProS', 'ProS','Ndnf HPF'\n",
    " 'NP PPP','NP SUB']\n",
    "\n",
    "non_neuron_list = [ 'VLMC', 'Oligo', 'Micro', 'Endo', 'Astro', 'CR', 'OPC', 'Peri',  'PVM', 'SMC', 'v3d']\n",
    "\n",
    "inhib_list =  ['Lamp5 Lhx6','Lamp5','Meis2','Pvalb Vipr2','Pvalb','Sncg','Sst Chodl','Sst','Vip']\n",
    "\n",
    "all_list = excite_list + non_neuron_list + inhib_list\n",
    "\n",
    "#add cell type data to frame\n",
    "mega_grps = []\n",
    "for i in mouse_means0[\"higher_cluster\"]:\n",
    "    if i in excite_list :\n",
    "        mega_grps.append(\"Excitatory\")\n",
    "    if i in inhib_list :\n",
    "        mega_grps.append(\"Inhibitory\")\n",
    "    if i in non_neuron_list :\n",
    "        mega_grps.append(\"Non-Neuronal\")\n",
    "    if i not in all_list :\n",
    "        mega_grps.append(\"Unknown\")\n",
    "mouse_means0[\"Cell_Type\"] = mega_grps\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "cells =[]\n",
    "genes = []\n",
    "graph_out1 = widgets.Output()  \n",
    "graph_out2 = widgets.Output()  \n",
    "def on_go_button_clicked(go_button):\n",
    "\n",
    "#Mark errors\n",
    "    if len(genes) < 1:\n",
    "        with outerrors:\n",
    "            print(\"Choose at least one Gene\")\n",
    "    if len(list(Cells_drop.value)) < 1:\n",
    "        with outerrors:\n",
    "            print(\"Choose at least one Cell Cluster\")\n",
    "            \n",
    "## restructure data frame to include selected types only \n",
    "    if len(genes) >= 1:\n",
    "         if len(list(Cells_drop.value)) >= 1:\n",
    "            data_gene_cell = mouse_means0[mouse_means0[\"higher_cluster\"].isin(list(Cells_drop.value))]\n",
    "            genes.append(\"cluster_label\")\n",
    "            genes.append(\"higher_cluster\")\n",
    "            genes.append(\"Cell_Type\")\n",
    "\n",
    "            data_gene_cell2 = data_gene_cell[genes]\n",
    "            data_gene_cell2 = data_gene_cell.reset_index()\n",
    "                        #data_gene_cell2\n",
    "            mouse_means = pd.melt(data_gene_cell2, id_vars =['cluster_label', 'higher_cluster','Cell_Type'],\n",
    "                                          var_name ='GeneID', value_name ='Expression'\n",
    "                                     )\n",
    "            \n",
    "            \n",
    "            \n",
    "   ###Graph it         \n",
    "            import plotly.graph_objects as go\n",
    "            fig = px.box(mouse_means, x=\"Cell_Type\", y=\"Expression\", color = \"Cell_Type\", \n",
    "                    facet_col = \"GeneID\",\n",
    "                    hover_name = \"Cell_Type\",\n",
    "                    hover_data = [\"Cell_Type\", \"cluster_label\"]\n",
    "                    )\n",
    "            #fig.update_layout(barmode='group')\n",
    "            with graph_out1:\n",
    "                    fig.show()\n",
    "\n",
    "            ## print table\n",
    "            fig2 = go.Figure(data=[go.Table(\n",
    "            header=dict(values=list(data_gene_cell2.columns),\n",
    "                    fill_color='paleturquoise',\n",
    "                    align='left'),\n",
    "            cells=dict(values=[data_gene_cell2.GeneID, data_gene_cell2.Cluster, data_gene_cell2.Expression,\n",
    "                              data_gene_cell2.Cell_Type, data_gene_cell2.Age, data_gene_cell2.Category],\n",
    "                    fill_color='lavender',\n",
    "                    align='left'))\n",
    "                ])\n",
    "            with graph_out2:\n",
    "                    fig2.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "##define GUI  \n",
    "def choose_gene(change):\n",
    "    if dropdown.value not in genes:\n",
    "        if dropdown.value != '':\n",
    "            genes.append(dropdown.value)\n",
    "            with outgenes:\n",
    "                display(dropdown.value)\n",
    "                \n",
    "                \n",
    "spacer = widgets.Label('')\n",
    "genetitle = widgets.Label('Select Genes')\n",
    "\n",
    "dropdown =  widgets.Combobox(\n",
    "    options= mouse_idx,\n",
    "    description='Gene',\n",
    "    placeholder='Type here',\n",
    "    layout={'width': '80%'})\n",
    "\n",
    "   \n",
    "outgenes = widgets.Output()   \n",
    "outgraph = widgets.Output(layout={'border': '1px solid black'})  \n",
    "\n",
    "        \n",
    "gene_button = widgets.Button(description = 'Enter Gene', \n",
    "                button_style = '', \n",
    "                display='flex',\n",
    "                flex_flow='column',\n",
    "                align_items='stretch')\n",
    "\n",
    "gene_button.on_click(choose_gene)\n",
    "\n",
    "celllabel = widgets.Label('Cell Clusters:')\n",
    "Cells_drop =widgets.SelectMultiple(\n",
    "    options=all_list,\n",
    "    rows=8,\n",
    "    layout={'width': 'max-content'}) \n",
    "\n",
    "\n",
    "go_button = widgets.Button(description = 'Produce Output', \n",
    "                           button_style = '', \n",
    "                            display='flex',\n",
    "                            flex_flow='column',\n",
    "                            align_items='stretch')\n",
    "go_button.on_click(on_go_button_clicked)\n",
    "\n",
    "gene_button.observe(dropdown.value)\n",
    "go_button.observe(dropdown.value)\n",
    "go_button.observe(Cells_drop.value)\n",
    "\n",
    "\n",
    "graph_out1 = widgets.Output()  \n",
    "graph_out2 = widgets.Output()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9b4a32942013491799227a435625a3c4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(HBox(children=(VBox(children=(Label(value='Cell Clusters:'), SelectMultiple(layout=Layout(widthâ€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genesbox = widgets.VBox([genetitle,dropdown, gene_button])\n",
    "cellsbox = widgets.VBox([celllabel, Cells_drop])\n",
    "genechoices = widgets.VBox([genesbox, outgenes,go_button])\n",
    "outerrors = widgets.Output() \n",
    "\n",
    "graphit = widgets.VBox([graph_out1, graph_out2])\n",
    "top_app = widgets.HBox([cellsbox,genechoices, outerrors])\n",
    "app = widgets.VBox([top_app, graphit])\n",
    "\n",
    "\n",
    "app"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hideCode": false,
    "hidePrompt": false,
    "hide_input": false
   },
   "outputs": [],
   "source": [
    "##define GUI\n",
    "table_style = {'description_width': 'initial'}\n",
    "useidx = ('choose species','then genes will appear')\n",
    "spacer = widgets.Label('')\n",
    "\n",
    "from threading import Timer\n",
    "\n",
    "from traitlets import traitlets\n",
    "\n",
    "class LoadedButton(widgets.Button):\n",
    "    \"\"\"A button that can hold a value as a attribute.\"\"\"\n",
    "\n",
    "    def __init__(self, value=None, *args, **kwargs):\n",
    "        super(LoadedButton, self).__init__(*args, **kwargs)\n",
    "        # Create the value attribute.\n",
    "        self.add_traits(value=traitlets.Any(value))\n",
    "#####################################################        \n",
    "genes = []\n",
    "genetitle = widgets.Label('Select Genes')\n",
    "dropdown =  widgets.Combobox(\n",
    "    options= mouse_idx,\n",
    "    description='Gene',\n",
    "    placeholder='Type here',\n",
    "    layout={'width': '80%'})\n",
    "\n",
    "def delay():\n",
    "    gene_button.button_style = ''\n",
    "    gene_button.disabled = False\n",
    "    \n",
    "def choose_gene(change):\n",
    "    if len(genes)<=7:\n",
    "        gene_button.button_style = 'success'\n",
    "        gene_button.disabled = True\n",
    "        genes.append(dropdown.value)     \n",
    "        d = Timer(0.5,delay)\n",
    "        d.start()\n",
    "    if len(genes)>7:       \n",
    "        gene_button.button_style = 'warning'\n",
    "        gene_button.description = \"Max Genes Selected\"\n",
    "        gene_button.disabled = True\n",
    "        \n",
    "gene_button = widgets.Button(description = 'Enter Gene', button_style = '')\n",
    "gene_button.on_click(choose_gene)\n",
    "\n",
    "\n",
    "choicelabel = widgets.Label('Analyze species:')\n",
    "msCells = widgets.Button(\n",
    "    description='Mouse',\n",
    "    disabled=False,\n",
    "    button_style='', # 'success', 'info', 'warning', 'danger' or ''\n",
    "    tooltip='Description',\n",
    "    width = '110%')\n",
    "\n",
    "huCells = widgets.Button(\n",
    "    description='Human',\n",
    "    disabled=False,\n",
    "    button_style='', # 'success', 'info', 'warning', 'danger' or ''\n",
    "    tooltip='Description',\n",
    "    width = '110%')\n",
    "\n",
    "#####################################################\n",
    "mscelllabel = widgets.Label('Mouse Cell Clusters:')\n",
    "msCells_drop =widgets.SelectMultiple(\n",
    "    options=mouse_listOfColumnNames,\n",
    "    disabled=True,\n",
    "    rows=17,\n",
    "    layout={'width': 'max-content'},\n",
    "    #style=table_style\n",
    ") \n",
    "\n",
    "hucelllabel = widgets.Label('Human Cell Clusters:')\n",
    "huCells_drop =widgets.SelectMultiple(\n",
    "    options=human_listOfColumnNames,\n",
    "    disabled=True,\n",
    "    rows=17,\n",
    "    layout={'width': 'max-content'},\n",
    "    #style=table_style\n",
    ") \n",
    "\n",
    "def choose_mouse(msCells):\n",
    "    huCells.disabled = True\n",
    "    msCells.button_style = 'success'\n",
    "    msCells_drop.disabled = False\n",
    "    huCells_drop.disabled = True \n",
    "    msCells.value = \"mouse\"\n",
    "\n",
    "msCells.on_click(choose_mouse)\n",
    "\n",
    "def choose_human(huCells):\n",
    "    huCells.button_style = 'success'\n",
    "    msCells.disabled = True\n",
    "    msCells_drop.disabled = True\n",
    "    huCells_drop.disabled = False\n",
    "    huCells.value = \"human\"\n",
    "    dropdown.options = human_idx\n",
    "\n",
    "huCells.on_click(choose_human)\n",
    "\n",
    "msCells = LoadedButton(description=\"Mouse\", value=\"not mouse\")\n",
    "msCells.on_click(choose_mouse)   \n",
    "\n",
    "huCells = LoadedButton(description=\"Human\", value=\"not human\")\n",
    "huCells.on_click(choose_human) \n",
    "dropdown.observe(choose_human, 'change')\n",
    "########################################\n",
    "\n",
    "go_button = widgets.Button(description = 'Produce Output', \n",
    "                           button_style = 'success', \n",
    "                            display='flex',\n",
    "                            flex_flow='column',\n",
    "                            align_items='stretch')\n",
    "reset_button = widgets.Button(description = 'Reset Options',\n",
    "                              button_style = 'warning',\n",
    "                            display='flex',\n",
    "                            flex_flow='column',\n",
    "                            align_items='stretch',\n",
    "                             width = 'auto')\n",
    "\n",
    "from ipywidgets import Widget\n",
    "cells =[]\n",
    "def on_go_button_clicked(go_button):\n",
    "    if huCells.value == 'human':\n",
    "        cells.append(huCells_drop.value)\n",
    "    else:\n",
    "        cells.append(msCells_drop.value)\n",
    "    expgraph(genes,cells)\n",
    "\n",
    "\n",
    "def on_reset_button_clicked(reset_button):  \n",
    "    Widget.close_all()\n",
    "    \n",
    "go_button.observe(huCells, 'value')   \n",
    "go_button.on_click(on_go_button_clicked)\n",
    "reset_button.on_click(on_reset_button_clicked)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "specieschoices = widgets.VBox([choicelabel,msCells, mscelllabel, msCells_drop])\n",
    "spacerhuman = widgets.VBox([spacer, huCells,hucelllabel, huCells_drop])\n",
    "genesbox = widgets.VBox([genetitle,dropdown, gene_button])\n",
    "buttons = widgets.HBox([go_button, spacer, reset_button])\n",
    "graphchoices = widgets.VBox([spacer, spacer, genesbox,spacer, spacer, spacer, buttons])\n",
    "#graphs2 = widgets.HBox([graphchoices])\n",
    "####################################################\n",
    "display(\"..............\")\n",
    "display(\"Choose your Species and Graph Type. For Expression level graphs, choose cell types to include. For Venn diagrams, all cell types will be included. Choose genes, pressing Enter after each choice. To produce graphs, press the green Produce Output button.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hideCode": false,
    "hidePrompt": false,
    "hide_input": true
   },
   "outputs": [],
   "source": [
    "%matplotlib notebook\n",
    "\n",
    "def expgraph(genes, cells): \n",
    "    if msCells.value == 'mouse':\n",
    "        bigframe = mouse_means0\n",
    "        cells = msCells_drop.value\n",
    "    if huCells.value == 'human':\n",
    "        bigframe = human_means\n",
    "        cells = huCells_drop.value\n",
    "\n",
    "    newdf = pd.DataFrame(columns = ['Cells','Genes','Expression'])\n",
    "    newdf['Cells'] = cells * len(genes)\n",
    "    print(newdf)\n",
    "\n",
    "    genelist=[]\n",
    "    for i in range(len(genes)):\n",
    "        genea = [genes[i]]*len(cells)\n",
    "        genelist.append(genea)\n",
    "    flat_geneslist = [item for sublist in genelist for item in sublist]\n",
    "    newdf['Genes'] = flat_geneslist\n",
    "\n",
    "    expressions =[]\n",
    "    for i in range(len(genes)*len(cells)):\n",
    "        value1 = bigframe.loc[newdf['Genes'][i], newdf['Cells'][i]]\n",
    "        expressions.append(value1)\n",
    "    newdf['Expression'] = expressions\n",
    "\n",
    "    if len(newdf['Expression'])>0:\n",
    "        maxm = max(newdf[\"Expression\"])\n",
    "    \n",
    "\n",
    "    #newdf.to_excel(excelname.value) \n",
    "    \n",
    "    sns.set_palette(\"colorblind\", 10)\n",
    "    g = sns.catplot(x=\"Genes\", y=\"Expression\", col=\"Cells\",\n",
    "                    data=newdf, \n",
    "                    saturation=1,\n",
    "                    kind=\"bar\", \n",
    "                    ci=None, \n",
    "                    aspect=.6, \n",
    "                    edgecolor=\"k\", \n",
    "                    linewidth=1\n",
    "                    )\n",
    "\n",
    "\n",
    "    (g.set_axis_labels(\"\", \"Expression\")\n",
    "          .set_xticklabels(genes)\n",
    "          .set_titles(\"{col_name}\")\n",
    "          .set(ylim=(0, (maxm + 2)))\n",
    "          .despine(left=True))  \n",
    "\n",
    "    for ax in g.axes.ravel():\n",
    "            ax.set_xticklabels(ax.get_xticklabels(), rotation=45, horizontalalignment='right')\n",
    "            for p in ax.patches:\n",
    "                ax.annotate('%.2f' % p.get_height(), (p.get_x() + p.get_width() / 2., p.get_height()), \n",
    "\n",
    "                            ha='center', va='center', xytext=(0, 8), \n",
    "                            textcoords='offset points')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hideCode": false,
    "hidePrompt": false,
    "hide_input": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### cell #1\n",
    "\n",
    "box_layout = widgets.Layout(\n",
    "        border='solid 1px black',\n",
    "        margin='0px 10px 10px 0px',\n",
    "        padding='5px 5px 5px 5px',\n",
    "        height='auto', width='auto')\n",
    "\n",
    "grid = GridspecLayout(1, 3)\n",
    "grid.layout = box_layout\n",
    "# fill it in with widgets\n",
    "grid[0,0] = specieschoices\n",
    "grid[0,1] = spacerhuman\n",
    "grid[0, 2:] = graphs2\n",
    "grid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hideCode": false,
    "hidePrompt": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_code_all_hidden": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
